File: Two_Pointers/TrappingRainWater/trappingRainWater.py
from typing import List


class Solution:
    #  UnOptimized but brute force solution
    def trap_UnOptimized(self, height: List[int]) -> int:
        n = len(height)
        maxLeftArray = [0]*n
        maxRightArray = [0]*n
        maxLeft = 0
        maxRight = 0
        minItem = 0
        sum_water = 0
        # we store into the maxleft array
        for i in range(n):
            maxLeftArray[i] = maxLeft
            maxLeft = max(maxLeft,height[i])
        # we store into the maxRight array
        for i in range(n-1, -1, -1):
            maxRightArray[i] = maxRight
            maxRight = max(maxRight, height[i])
        # we store into the minArray
        for i in range(n):
            minItem = min(maxLeftArray[i], maxRightArray[i])
            relativeWater = minItem - height[i]
            sum_water += relativeWater if relativeWater > 0 else 0
        return sum_water
            
    def trap(self, height: List[int]) -> int:
        if not height:
            return 0
        l = 0
        r = len(height)-1
        leftMax = height[0]
        rightMax = height[-1]
        res = 0
        while l < r:
            if leftMax < rightMax:
                l += 1
                leftMax = max(leftMax, height[l])
                res += leftMax - height[l]
            else:
                r -= 1
                rightMax = max(rightMax, height[r])
                res += rightMax - height[r]
        return res

File: Two_Pointers/TrappingRainWater/TrappingRainWater.java
package Two_Pointers.TrappingRainWater;

class TrappingRainWater {
    public int trap(int[] height) {
        // Brute force approach:
        // Create a max array for the left iteration of the rain water
        // Create another max array for the right iteration
        // Now go through both the arrays and add to the sum += min(left_arr[i], right_arr[i])-height[i]
        // Return the sum
        int n = height.length, rainWater = 0;
        int[] leftMaxArray = new int[n];
        int[] rightMaxArray = new int[n];
        // Populate the leftMaxArray
        for (int i=1; i < n; i++){
            leftMaxArray[i] = Math.max(leftMaxArray[i-1], height[i-1]);
        }
        // Populate the rightMaxArray
        for (int i=n-1; i > 0; i--){
            rightMaxArray[i-1] = Math.max(rightMaxArray[i], height[i]);
        }
        // Now go through both the arrays and add to the sum += min(left_arr[i], right_arr[i])-height[i]
        for (int i=0; i<n; i++){
            int currentVal = Math.min(leftMaxArray[i], rightMaxArray[i])-height[i];
            rainWater += currentVal > 0 ? currentVal : 0;
        }
        return rainWater;
    }
    public int trap_optimized(int[] height) {
        // Optimized version: We use a Two Ptr approach starting from left and Right Ptr 
        // Check if the leftMax is lesser than the rightMax
        // Then we increment l as we start it 
        //  sub leftMax = Math.min(leftMax, height[l]);
        //  add the leftMax - height[l] to the rainwater count; 
        // Else we increment r as we start it 
        //  sub rightMax = Math.min(rightMax, height[r]);
        //  add the rightMax - height[r] to the rainwater count;
        int l =0, r = height.length-1, result = 0;
        int maxLeft=height[l], maxRight=height[r];
        while(l < r) {
            if (maxLeft < maxRight){
                l++;
                maxLeft = Math.max(maxLeft, height[l]);
                result += maxLeft-height[l];
            } else {
                r--;
                maxRight = Math.max(maxRight, height[r]);
                result += maxRight-height[r];
            }
        }
        return result;
    }
}